File: singularity_economy/companies.py
"""
Company universe: maps each candidate trade to model pools with analytic
judgments (pool weights, operating-leverage beta, share drift, terminal
multiple). Market caps / NTM earnings are placeholders from the research sweep
until `load_financials()` overrides them with the phase-2 deep-dive output.

Pool-mapping conventions
------------------------
- `pools` weights sum to ~1.0. `gdp_index` is the neutral "grows with the
  economy" bucket for conglomerate/legacy segments.
- `pool_beta` >1: operating leverage or price-taker torque on the pool;
  <1: contracted/backlogged smoothing.
- `share_drift`: annual share gain(+)/loss(-) vs the pool — company-specific
  judgments (socket losses, commoditization, capacity-led share gains).
- Consumer-credit and duration trades (SYF, ALLY, TLT) are NOT pool-mappable;
  they are evaluated qualitatively in the report (transition-recession tail).
"""

from __future__ import annotations

import json

from valuation import Company

# stance, mcap_b, ntm_earnings_b are refreshed from phase-2 financials.
COMPANIES: list[Company] = [
    # ---------------- LONGS: compute complex ----------------
    Company("TSM", "TSMC", 1930, 100, {"silicon": 0.9, "gdp_index": 0.1},
            pool_beta=0.9, share_drift=0.01, terminal_multiple=18, stance="long",
            notes="Dual chokepoint (N2 + CoWoS); IP-moat rent class -> durable"),
    Company("NVDA", "NVIDIA", 5060, 240, {"silicon": 1.0},
            pool_beta=1.15, share_drift=-0.03, terminal_multiple=16, stance="long",
            notes="Share loss to ASICs inside a pool growing 2-4x consensus"),
    Company("AVGO", "Broadcom", 1870, 75, {"silicon": 0.8, "gdp_index": 0.2},
            pool_beta=1.05, share_drift=0.03, terminal_multiple=20, stance="long",
            notes="Custom-ASIC royalty on hyperscaler in-sourcing"),
    Company("MU", "Micron", 959, 120, {"silicon": 1.0},
            pool_beta=0.6, share_drift=0.01, terminal_multiple=8, stance="long",
            notes="Capacity-scarcity rent class -> harvest early; low terminal multiple"),
    Company("000660.KS", "SK Hynix", 1200, 170, {"silicon": 1.0},
            pool_beta=0.6, share_drift=0.0, terminal_multiple=8, stance="long",
            notes="HBM leader; same cycle caveat as MU"),
    Company("BESI.AS", "BE Semiconductor", 15, 0.35, {"silicon": 1.0},
            pool_beta=1.3, share_drift=0.04, terminal_multiple=22, stance="long",
            notes="Hybrid bonding adoption = share gain against stacking intensity"),
    Company("ASML", "ASML", 450, 12.5, {"silicon": 1.0},
            pool_beta=0.9, share_drift=0.01, terminal_multiple=22, stance="long",
            notes="Litho monopoly; IP-moat rent class"),
    # ---------------- LONGS: power / grid ----------------
    Company("GEV", "GE Vernova", 274, 8.5, {"power_equipment": 1.0},
            pool_beta=1.1, share_drift=0.01, terminal_multiple=18, stance="long",
            notes="Sold out through 2030; backlog = contracted scarcity"),
    Company("VST", "Vistra", 65, 3.8, {"gdp_index": 1.0},
            pool_beta=0.4, share_drift=0.0, terminal_multiple=14, stance="long",
            new_pool_capture={"electricity": (0.05, 0.5)},
            notes="Merchant torque to power scarcity at a discount to CEG"),
    Company("NRG", "NRG Energy", 35, 2.4, {"gdp_index": 1.0},
            pool_beta=0.4, share_drift=0.01, terminal_multiple=13, stance="long",
            new_pool_capture={"electricity": (0.045, 0.45)}),
    Company("CEG", "Constellation", 110, 3.6, {"gdp_index": 1.0},
            pool_beta=0.4, share_drift=0.01, terminal_multiple=18, stance="long",
            new_pool_capture={"electricity": (0.06, 0.5)},
            notes="Contracted 20yr PPAs = royalty-acre economics"),
    Company("POWL", "Powell Industries", 4.5, 0.19, {"power_equipment": 1.0},
            pool_beta=1.3, share_drift=0.01, terminal_multiple=15, stance="long"),
    Company("6501.T", "Hitachi", 130, 6.0,
            {"power_equipment": 0.45, "dc_infra": 0.15, "gdp_index": 0.40},
            pool_beta=1.0, share_drift=0.005, terminal_multiple=15, stance="long",
            notes="Hitachi Energy transformer bottleneck inside a conglomerate"),
    # ---------------- LONGS: networking / optics ----------------
    Company("LITE", "Lumentum", 12, 0.45, {"silicon": 1.0},
            pool_beta=1.2, share_drift=0.02, terminal_multiple=14, stance="long",
            notes="EML sold out past 2027; capacity-scarcity class -> monitor supply response"),
    Company("CLS", "Celestica", 22, 0.65, {"dc_infra": 0.6, "silicon": 0.4},
            pool_beta=1.1, share_drift=0.02, terminal_multiple=15, stance="long"),
    # ---------------- LONGS: robotics chain ----------------
    Company("MP", "MP Materials", 8, 0.05, {"gdp_index": 1.0},
            pool_beta=1.0, share_drift=0.10, terminal_multiple=20, stance="long",
            new_pool_capture={"robot_components": (0.10, 0.25)},
            notes="Geopolitical scarcity (non-China magnets), DoD price floor"),
    Company("6268.T", "Nabtesco", 4, 0.15, {"gdp_index": 1.0},
            pool_beta=1.0, share_drift=0.0, terminal_multiple=16, stance="long",
            new_pool_capture={"robot_components": (0.08, 0.22)}),
    Company("SHA.DE", "Schaeffler", 6, 0.7, {"gdp_index": 1.0},
            pool_beta=1.0, share_drift=-0.01, terminal_multiple=9, stance="long",
            new_pool_capture={"robot_components": (0.05, 0.15)},
            notes="Deep value + robot components optionality; ICE drag real"),
    Company("SYM", "Symbotic", 26, 0.35, {"gdp_index": 1.0},
            pool_beta=1.2, share_drift=0.10, terminal_multiple=22, stance="long",
            new_pool_capture={"robot_services": (0.03, 0.25)},
            notes="Contracted backlog converts ahead of humanoid hype"),
    # ---------------- LONGS: platforms / commodities / info ----------------
    Company("MSFT", "Microsoft", 3400, 118, {"seat_saas": 0.3, "gdp_index": 0.7},
            pool_beta=0.9, share_drift=0.01, terminal_multiple=18, stance="long",
            new_pool_capture={"ai_services": (0.12, 0.40)},
            notes="Aggregator: agent distribution offsets seat erosion"),
    Company("GOOGL", "Alphabet", 2400, 135, {"gdp_index": 0.85, "silicon": 0.15},
            pool_beta=1.0, share_drift=0.01, terminal_multiple=16, stance="long",
            new_pool_capture={"ai_services": (0.08, 0.40)},
            notes="TPU full stack + Waymo optionality at market multiple"),
    Company("FCX", "Freeport-McMoRan", 65, 4.2, {"gdp_index": 0.6, "dc_infra": 0.4},
            pool_beta=1.4, share_drift=0.0, terminal_multiple=12, stance="long",
            notes="Copper deficit proxy; beta carries the commodity torque"),
    Company("EQT", "EQT", 35, 2.2, {"gdp_index": 1.0},
            pool_beta=0.8, share_drift=0.0, terminal_multiple=12, stance="long",
            new_pool_capture={"electricity": (0.05, 0.35)}),
    Company("U-UN.TO", "Sprott Physical Uranium", 5.2, 0.0, {"gdp_index": 1.0},
            pool_beta=0.8, share_drift=0.0, terminal_multiple=1, stance="long",
            notes="NAV vehicle - valuation engine not meaningful; qualitative"),
    Company("WTKWY", "Wolters Kluwer", 30, 1.6, {"prof_info": 1.0},
            pool_beta=0.9, share_drift=0.02, terminal_multiple=15, stance="long",
            notes="Regulated-workflow moat inside a decaying pool class"),
    # ---------------- SHORTS ----------------
    Company("ADP", "ADP", 105, 4.3, {"human_cognitive_wages": 0.9, "gdp_index": 0.1},
            pool_beta=1.2, share_drift=0.02, terminal_multiple=14, stance="short",
            notes="Pays-per-control elasticity 0.7-1.0 to shrinking employment"),
    Company("PAYX", "Paychex", 48, 1.9, {"human_cognitive_wages": 0.9, "gdp_index": 0.1},
            pool_beta=1.2, share_drift=0.015, terminal_multiple=14, stance="short"),
    Company("RHI", "Robert Half", 4.5, 0.20, {"human_cognitive_wages": 1.0},
            pool_beta=2.0, share_drift=-0.02, terminal_multiple=9, stance="short",
            notes="Perm placement = first casualty; payout > NI"),
    Company("MAN", "ManpowerGroup", 3.5, 0.25, {"human_physical_wages": 0.6,
                                                "human_cognitive_wages": 0.4},
            pool_beta=1.8, share_drift=-0.02, terminal_multiple=8, stance="short",
            notes="Industrial staffing unpriced for 2028+ robotics"),
    Company("TCS.NS", "Tata Consultancy", 95, 6.0, {"it_services": 1.0},
            pool_beta=1.3, share_drift=0.0, terminal_multiple=10, stance="short",
            notes="Premium multiple vs peers with max labor-arbitrage exposure"),
    Company("WDAY", "Workday", 31.6, 2.4, {"seat_saas": 1.0},
            pool_beta=1.2, share_drift=0.0, terminal_multiple=11, stance="short"),
    Company("TEAM", "Atlassian", 20.3, 1.4, {"seat_saas": 1.0},
            pool_beta=1.3, share_drift=-0.02, terminal_multiple=11, stance="short",
            notes="Agent-native pipelines collapse the tracked-work loop"),
    Company("HUBS", "HubSpot", 9.7, 0.65, {"seat_saas": 1.0},
            pool_beta=1.3, share_drift=-0.02, terminal_multiple=11, stance="short"),
    Company("FDS", "FactSet", 12, 0.60, {"seat_saas": 0.5, "prof_info": 0.5},
            pool_beta=1.0, share_drift=-0.01, terminal_multiple=12, stance="short",
            notes="ASV tied to front-office headcount"),
    Company("MMC", "Marsh McLennan", 95, 4.6, {"prof_info": 0.8, "gdp_index": 0.2},
            pool_beta=1.0, share_drift=0.0, terminal_multiple=14, stance="short"),
    Company("CHRW", "CH Robinson", 18, 0.75, {"human_physical_wages": 0.5,
                                              "gdp_index": 0.5},
            pool_beta=1.0, share_drift=-0.05, terminal_multiple=11, stance="short",
            notes="AI-advantage narrative vs AI commoditizing brokerage itself"),
    Company("LSTR", "Landstar", 6, 0.30, {"human_physical_wages": 0.6, "gdp_index": 0.4},
            pool_beta=1.0, share_drift=-0.04, terminal_multiple=10, stance="short"),
    Company("MRVL", "Marvell", 200, 3.2, {"silicon": 1.0},
            pool_beta=1.0, share_drift=-0.08, terminal_multiple=13, stance="short",
            notes="Socket erosion inside the boom; pair vs AVGO long"),
    Company("6954.T", "FANUC", 44, 1.5, {"gdp_index": 1.0},
            pool_beta=0.9, share_drift=-0.02, terminal_multiple=18, stance="short",
            new_pool_capture={"robots": (0.05, 0.20)},
            notes="Humanoid halo at 39x with zero humanoid revenue"),
    Company("6324.T", "Harmonic Drive", 6, 0.02, {"gdp_index": 1.0},
            pool_beta=1.0, share_drift=0.0, terminal_multiple=25, stance="short",
            new_pool_capture={"robot_components": (0.06, 0.28)},
            notes="404x trailing; Corning-2000 setup vs Chinese commoditization"),
    Company("EQIX", "Equinix", 75, 2.6, {"dc_infra": 0.3, "gdp_index": 0.7},
            pool_beta=0.5, share_drift=-0.01, terminal_multiple=16, stance="short",
            notes="Margin squeeze between equipment suppliers and hyperscale tenants"),
    Company("PLTR", "Palantir", 320, 2.3, {"seat_saas": 0.3, "gdp_index": 0.7},
            pool_beta=1.1, share_drift=0.0, terminal_multiple=25, stance="short",
            new_pool_capture={"ai_services": (0.015, 0.30)},
            notes="139x trailing: growth is real, hurdle is impossible"),
]


def load_financials(path: str) -> list[Company]:
    """Override placeholder mcap/NTM/terminal-multiple with phase-2 agent data."""
    with open(path) as f:
        fin = json.load(f)
    out = []
    for c in COMPANIES:
        d = fin.get(c.ticker)
        if d:
            c.mcap_b = d.get("mcap_b", c.mcap_b)
            ntm = d.get("ntm_earnings_b", c.ntm_earnings_b)
            if ntm and ntm > 0:
                c.ntm_earnings_b = ntm
            tm = d.get("suggested_terminal_multiple")
            if tm and 4 <= tm <= 30:
                c.terminal_multiple = tm
        out.append(c)
    return out


if __name__ == "__main__":
    import sys
    from valuation import evaluate_all
    comps = COMPANIES
    if len(sys.argv) > 1:
        comps = load_financials(sys.argv[1])
    comps = [c for c in comps if c.ntm_earnings_b > 0]  # NAV vehicles handled separately
    rows = evaluate_all(comps)
    json.dump(rows, open("output/valuations.json", "w"), indent=1)
    hdr = f"{'ticker':10} {'stance':6} {'implied CAGR':>12} {'E[up]':>7} {'worst':>7} {'best':>7}"
    print(hdr); print("-" * len(hdr))
    for r in rows:
        print(f"{r['ticker']:10} {r['stance']:6} {r['implied_cagr_at_current_price']:>12.1%} "
              f"{r['expected_upside']:>7.1%} {r['worst_scenario_upside']:>7.1%} "
              f"{r['best_scenario_upside']:>7.1%}")

File: singularity_economy/rust/singularity-econ/src/companies.rs
//! Company universe (port of the retired `companies.py` — this file is now
//! the specification). Pool weights, betas, share drifts, and terminal
//! multiples are analytic judgments; mcap/NTM/terminal multiples are SYNCED
//! from output/financials.json (phase-2 verified) and can be re-refreshed at
//! runtime via `load_financials`.

use crate::valuation::{Company, EmergingPool as E, RatioPool as R, Stance};

pub fn universe() -> Vec<Company> {
    use Stance::*;
    let c = |ticker, name, mcap_b, ntm, pools, beta, drift, tm, stance, capture, notes| Company {
        ticker, name, mcap_b, ntm_earnings_b: ntm, pools, pool_beta: beta,
        share_drift: drift, terminal_multiple: tm, stance, capture,
        taiwan_fab_exposure: 0.0, notes,
    };
    let mut universe = vec![
        // ---- LONGS: compute complex ----
        c("TSM", "TSMC", 2184.0, 112.0, vec![(R::Silicon, 0.9), (R::GdpIndex, 0.1)],
          0.9, 0.01, 18.0, Long, vec![],
          "Dual chokepoint (N2 + CoWoS); IP-moat rent class"),
        c("NVDA", "NVIDIA", 5060.0, 241.0, vec![(R::Silicon, 1.0)],
          1.15, -0.03, 18.0, Long, vec![],
          "Share loss to ASICs inside a pool growing 2-4x consensus"),
        c("AVGO", "Broadcom", 1830.0, 75.0, vec![(R::Silicon, 0.8), (R::GdpIndex, 0.2)],
          1.05, 0.03, 19.0, Long, vec![],
          "Custom-ASIC royalty; round-2: vendor-financier drift, half-size"),
        c("MU", "Micron", 1120.0, 162.0, vec![(R::Silicon, 1.0)],
          0.6, 0.01, 10.0, Long, vec![],
          "Capacity-scarcity rent class -> harvest early (watchlist-gated)"),
        c("000660.KS", "SK Hynix", 923.0, 140.0, vec![(R::Silicon, 1.0)],
          0.6, 0.0, 8.5, Long, vec![],
          "HBM leader; same cycle caveat as MU (watchlist-gated)"),
        c("BESI.AS", "BE Semiconductor", 20.7, 0.51, vec![(R::Silicon, 1.0)],
          1.3, 0.03, 22.0, Long, vec![],
          "Hybrid bonding share gain; drift aligned with LITE (Sep-26)"),
        c("ASML", "ASML", 692.5, 14.0, vec![(R::Silicon, 1.0)],
          0.9, 0.01, 24.0, Long, vec![],
          "Litho monopoly; IP-moat rent class; scale in toward ~35x"),
        // ---- LONGS: power / grid ----
        c("GEV", "GE Vernova", 274.6, 5.67,
          vec![(R::PowerEquipment, 0.75), (R::GdpIndex, 0.25)],
          1.1, 0.01, 20.0, Long, vec![],
          "Gas turbines + grid; wind (~25%) structurally impaired (Sep-26 remap 0.75/0.25)"),
        c("VST", "Vistra", 57.0, 3.05, vec![(R::GdpIndex, 1.0)],
          0.4, 0.0, 13.0, Long, vec![(E::Electricity, 0.020)],
          "Merchant torque to power scarcity"),
        c("NRG", "NRG Energy", 30.2, 2.18, vec![(R::GdpIndex, 1.0)],
          0.4, 0.01, 12.0, Long, vec![(E::Electricity, 0.0065)],
          "13GW gas below replacement; size 50-60% of other power legs"),
        c("CEG", "Constellation", 99.3, 4.5, vec![(R::GdpIndex, 1.0)],
          0.4, 0.01, 17.0, Long, vec![(E::Electricity, 0.016)],
          "Contracted nuclear PPAs; round-2: half gas post-Calpine, half-size"),
        c("POWL", "Powell Industries", 8.83, 0.21,
          vec![(R::PowerEquipment, 0.55), (R::GdpIndex, 0.45)],
          1.3, 0.01, 15.0, Long, vec![],
          "Switchgear; ~45% oil/gas/petrochem (Sep-26 remap 0.55 PowerEq / 0.45 GDP)"),
        c("6501.T", "Hitachi", 147.0, 6.15,
          vec![(R::PowerEquipment, 0.45), (R::DcInfra, 0.15), (R::GdpIndex, 0.40)],
          1.0, 0.005, 18.0, Long, vec![],
          "Hitachi Energy transformer bottleneck inside a conglomerate"),
        // ---- LONGS: networking / optics ----
        c("LITE", "Lumentum", 64.9, 1.24, vec![(R::Silicon, 1.0)],
          1.2, 0.03, 15.0, Long, vec![],
          "EML scarcity; drift aligned with BESI's intensity-gain story (Sep-26)"),
        c("CLS", "Celestica", 38.4, 1.3, vec![(R::DcInfra, 0.8), (R::Silicon, 0.2)],
          1.0, 0.02, 14.0, Long, vec![],
          "Assembler: captures assembly value, not silicon rent (Sep-26 remap DcInfra 0.8/Si 0.2)"),
        // ---- LONGS: robotics chain (v2 flips these negative) ----
        c("MP", "MP Materials", 7.95, 0.05, vec![(R::GdpIndex, 1.0)],
          1.0, 0.10, 15.0, Long, vec![(E::RobotComponents, 0.10)],
          "Geopolitical magnet scarcity; v2 demotes to hold-only"),
        c("6268.T", "Nabtesco", 4.0, 0.155, vec![(R::GdpIndex, 1.0)],
          1.0, 0.0, 16.0, Long, vec![],
          "RV cycloidal reducers != humanoid joints: capture removed (Sep-26 category fix)"),
        c("SHA.DE", "Schaeffler", 8.68, 0.5, vec![(R::GdpIndex, 1.0)],
          1.0, -0.01, 10.0, Long, vec![(E::RobotComponents, 0.05)],
          "KILLED (humanoid revenue immaterial on horizon)"),
        // overnight components deep-dive: cheapest reasonably-valued chokepoints
        c("002472.SZ", "Shuanghuan", 4.1, 0.18, vec![(R::GdpIndex, 1.0)],
          1.0, 0.05, 24.0, Long, vec![(E::RobotComponents, 0.06)],
          "Cheapest reducer play (~24x, profitable RV arm; new 500k/yr Suzhou plant 2026)"),
        c("300748.SZ", "JL MAG", 4.2, 0.11, vec![(R::GdpIndex, 1.0)],
          1.1, 0.04, 36.0, Long, vec![(E::RobotComponents, 0.05)],
          "China NdFeB magnet benchmark; RE-motor feedstock, substitution risk two-sided"),
        c("TECK", "Teck Resources", 29.6, 1.76,
          vec![(R::GdpIndex, 0.75), (R::PowerEquipment, 0.15), (R::DcInfra, 0.10)],
          1.2, 0.0, 14.0, Long, vec![],
          "Copper: DCs are low-single-digit % of demand, grid more (Sep-26 remap); Anglo Teck merger pending SAMR"),
        c("SYM", "Symbotic", 24.45, 0.35, vec![(R::GdpIndex, 1.0)],
          1.2, 0.10, 22.0, Long, vec![],
          "Integrator, not a component chokepoint: demand is retailer capex (Sep-26: RobotServices capture removed)"),
        // ---- LONGS: platforms / commodities / info ----
        c("MSFT", "Microsoft", 2830.0, 137.0,
          vec![(R::SeatSaas, 0.3), (R::GdpIndex, 0.7)],
          0.9, 0.01, 22.0, Long, vec![(E::AiServices, 0.12)],
          "Aggregator; round-2: watch ex-OpenAI RPO growth, medium size"),
        c("GOOGL", "Alphabet", 3890.0, 175.0,
          vec![(R::GdpIndex, 0.55), (R::ProfInfo, 0.20), (R::DcInfra, 0.10), (R::Silicon, 0.15)],
          1.0, 0.01, 20.0, Long, vec![(E::AiServices, 0.08)],
          "Barbell: LLM-disruptable search rent (ProfInfo) + Cloud (DcInfra) + TPU (Si) (Sep-26 remap)"),
        c("FCX", "Freeport-McMoRan", 91.29, 4.6,
          vec![(R::GdpIndex, 0.75), (R::PowerEquipment, 0.15), (R::DcInfra, 0.10)],
          1.2, 0.0, 12.0, Long, vec![],
          "Copper proxy (killed at price; expression moved to deferred curve)"),
        c("EQT", "EQT", 33.4, 2.3, vec![(R::GdpIndex, 1.0)],
          0.8, 0.0, 10.0, Long, vec![],
          "Gas producer: sells molecules, not MWh — Electricity capture removed (Sep-26 sign fix); no gas-price pool yet"),
        c("WTKWY", "Wolters Kluwer", 15.6, 1.5, vec![(R::ProfInfo, 1.0)],
          0.9, 0.02, 12.0, Long, vec![],
          "No position: model/red-team disagreement"),
        // ---- SHORTS ----
        c("ADP", "ADP", 97.1, 4.8,
          vec![(R::HumanCognitiveWages, 0.9), (R::GdpIndex, 0.1)],
          0.6, 0.02, 16.0, Short, vec![],
          "Pays-per-control elasticity <0.5 empirically: beta 1.2->0.6 (Sep-26)"),
        c("PAYX", "Paychex", 39.4, 2.1,
          vec![(R::HumanPhysicalWages, 0.55), (R::HumanCognitiveWages, 0.35), (R::GdpIndex, 0.1)],
          1.0, 0.015, 15.0, Short, vec![],
          "SMB base is physical-labor heavy (Sep-26 remap: majority HumanPhysicalWages, beta 1.0)"),
        c("RHI", "Robert Half", 3.6, 0.16, vec![(R::HumanCognitiveWages, 1.0)],
          2.0, -0.02, 8.0, Short, vec![],
          "Live short via 4-7mo put-spread ladders (round-2 restructure)"),
        c("MAN", "ManpowerGroup", 2.43, 0.17,
          vec![(R::HumanPhysicalWages, 0.6), (R::HumanCognitiveWages, 0.4)],
          1.8, -0.02, 8.0, Short, vec![],
          "KILLED (19.9% SI squeeze trap)"),
        c("TCS.NS", "Tata Consultancy", 97.5, 6.5, vec![(R::ItServices, 1.0)],
          1.3, 0.0, 11.0, Short, vec![],
          "KILLED high conf (thesis-incoherent: they sell the integration)"),
        c("WDAY", "Workday", 31.6, 2.9, vec![(R::SeatSaas, 1.0)],
          1.2, 0.0, 13.0, Short, vec![],
          "KILLED as delta-one; trigger-gated put spreads (cRPO <10%)"),
        c("TEAM", "Atlassian", 20.3, 1.51, vec![(R::SeatSaas, 1.0)],
          1.3, -0.02, 14.0, Short, vec![],
          "KILLED high conf (capture-ratio category error)"),
        c("HUBS", "HubSpot", 9.7, 0.72, vec![(R::SeatSaas, 1.0)],
          1.3, -0.02, 13.0, Short, vec![],
          "Live short, small, kill-watch (outcome-pricing pivot)"),
        c("FDS", "FactSet", 8.68, 0.68,
          vec![(R::SeatSaas, 0.5), (R::ProfInfo, 0.5)],
          1.0, -0.01, 10.0, Short, vec![],
          "KILLED by red team"),
        c("MMC", "Marsh McLennan", 84.2, 5.24,
          vec![(R::ProfInfo, 0.25), (R::GdpIndex, 0.75)],
          1.0, 0.0, 14.0, Short, vec![],
          "Broking is premium-linked (GDP); only Mercer/OW is ProfInfo (Sep-26 remap)"),
        c("CHRW", "CH Robinson", 24.2, 0.76,
          vec![(R::HumanPhysicalWages, 0.5), (R::GdpIndex, 0.5)],
          1.0, -0.05, 12.0, Short, vec![],
          "Live short post-print ($210-225 entry)"),
        c("LSTR", "Landstar", 7.06, 0.21,
          vec![(R::HumanPhysicalWages, 0.5), (R::GdpIndex, 0.3), (R::DcInfra, 0.2)],
          1.0, -0.04, 11.0, Short, vec![],
          "Freight arm of the AI buildout: +0.2 DcInfra (Sep-26)"),
        c("MRVL", "Marvell", 183.3, 3.98, vec![(R::Silicon, 1.0)],
          1.0, -0.05, 22.0, Short, vec![],
          "ASIC share loss partly offset by electro-optics: drift -0.08->-0.05 (Sep-26)"),
        c("6954.T", "FANUC", 44.0, 1.26, vec![(R::GdpIndex, 1.0)],
          0.9, -0.02, 22.0, Short, vec![(E::Robots, 0.10)],
          "World's largest industrial-robot maker: Robots capture 0.05->0.10 (Sep-26)"),
        c("6324.T", "Harmonic Drive", 4.43, 0.031, vec![(R::GdpIndex, 1.0)],
          1.0, 0.0, 18.0, Short, vec![],
          "Harmonic reducers; RobotComponents accrues to Chinese entrants — capture removed (Sep-26)"),
        c("EQIX", "Equinix", 101.9, 4.24,
          vec![(R::DcInfra, 0.3), (R::GdpIndex, 0.7)],
          0.5, -0.01, 18.0, Short, vec![],
          "KILLED by red team"),
        c("PLTR", "Palantir", 323.3, 4.0,
          vec![(R::SeatSaas, 0.3), (R::GdpIndex, 0.7)],
          1.1, 0.0, 27.0, Short, vec![(E::AiServices, 0.015)],
          "Relative short vs MSFT/GOOGL sleeve; options only post-Aug-3"),
    ];
    // Taiwan-fab exposure: TSMC's leading-edge capacity is overwhelmingly on-island
    // (Arizona/Japan a small, growing share), so an invasion destroys most of it —
    // a company-specific hit the global Silicon pool (whose scarcity margin RISES)
    // cannot express. Fabless names take the volume hit through the pool, exposure 0.
    if let Some(tsm) = universe.iter_mut().find(|c| c.ticker == "TSM") {
        tsm.taiwan_fab_exposure = 0.78; // Sep-26: ~0.85 today, overseas fabs ramping; value at the 2028 shock
    }
    universe
}

/// Override mcap / NTM / terminal multiple from the phase-2 verified
/// financials JSON ({ticker: {mcap_b, ntm_earnings_b,
/// suggested_terminal_multiple, ...}}).
#[cfg(feature = "serde")]
pub fn load_financials(companies: &mut [Company], json: &str) {
    let parsed: serde_json::Value = serde_json::from_str(json)
        .unwrap_or_else(|e| panic!("malformed financials JSON: {e}"));
    for c in companies.iter_mut() {
        if let Some(d) = parsed.get(c.ticker) {
            if let Some(m) = d.get("mcap_b").and_then(|v| v.as_f64()) {
                c.mcap_b = m;
            }
            if let Some(e) = d.get("ntm_earnings_b").and_then(|v| v.as_f64()) {
                if e > 0.0 {
                    c.ntm_earnings_b = e;
                }
            }
            if let Some(t) = d.get("suggested_terminal_multiple")
                .and_then(|v| v.as_f64())
            {
                if (4.0..=30.0).contains(&t) {
                    c.terminal_multiple = t;
                }
            }
        }
    }
}
